### ISSUE TITLE
[Core Dev]: Dynamic binding and the External Repository.

### ISSUE THREAD

**Issue by KTSnowy** [otterkit/otterkit-cobol#15]
### The Standard COBOL External Repository

Standard COBOL's External Repository contains all information required for activating (calling/invoking) programs, functions, or methods and for checking conformance (Checking their signature, in case dynamic binding is required).

The information contained in the repository is:
- The name of the source unit.
- The type of the source unit (Whether it is a program, function, class, or an interface).
- A description of its parameters (Their type declarations, whether they are optional or not, and if they are passed by value or by reference. Classes only: generics and inheritance).
- A description of the returning data item (and its type declaration).
- The exceptions that could be raised in the source unit.
- The object properties and methods (Classes only: the previous items also applies to store its methods).
- Extra information such as: if `DECIMAL-POINT IS COMMA` clause is specified, currency symbols, and locale information.
- Also, "Any other information that the implementation requires".

The Standard requires us to "provide a mechanism that allows the user to specify whether to update the external repository when a compilation unit is compiled".

From this we can tell that the external repository works like a local manifest which contains all program, function, class, method and interface signatures that belong to a given project. It can also be used during runtime when the signature of a program can't be statically checked during compile time (Raises an exception if it fails to find the signature at runtime).

If updating the external repository is turned off by the user then any kind of runtime signature check will fail if the signature was not added to it before the user turned it off.

```
*> External repository is global to each project
*> It cannot contain duplicate or conflicting program, function and class names


           ------------- Project dependency signatures
           |
           ˅
External Repository  <------- User project signatures
           ˄
           |
           ------------- Extra library signatures

```

The requirement to allow the user to update the repository tells us that it either has to be a local manifest file (probably too slow), or a compile time source-generated data structure populated with all signatures from a project (Including dependencies), with a compiler option to not update it.

### The problem with dynamic binding and the External Repository

There's a certain problem with storing all of this information in a single place, is that there's no way to store and return different delegate types from a single data structure that allows searching by a string value like a dictionary (I might be wrong about this).

I thought about storing the types themselves with `System.Type`, but that would require heavy use of reflection to get them back and call the procedure method from it.

```csharp
// Dictionaries *have* to contain the type of delegate it will store
// But we can't store just a single *kind* of signature
var dict = new Dictionary<string, Func<ICOBOLType, ICOBOLType>>();

// Something like this would not be possible (I think):
var signatures = new Dictionary<string, Signature<type or void, any params...>>();
```

A source-generated static class with different fields or properties to store those might work, but only if it still allows lookup with an arbitrary user-defined string value without causing a compiler error (I'm not sure how exactly that would look like).

```cobol
*> This is because COBOL allows getting a program pointer from a string:
set ptr-name to address of program str-var.

*> The value of str-var can't be statically checked during 
*> compile time if it is obtained from user input:
accept str-var from standard-input.

*> The name of the program will be checked during runtime
*> and allows calling it afterwards, which will cause another
*> runtime check for its parameters and return types:
call ptr-name using ... returning ... .
```

Dynamic binding like this can't be done without reflection in C# as far as I know.

```csharp
// This is exactly what we need, but it requires reflection:
Type.GetType("TypeName").GetMethod("MethodName");
```

Information about C# types and signatures might be available in the compiled assemblies or during compilation, but I'm not sure how to access those with Otterkit without using reflection (and how to use those during runtime to call methods). 

We're trying to avoid the use of reflection in order to support NativeAOT, with both the compiler and the generated C# code. Which probably means that this will have to be source-generated, ideally without causing a C# compiler error or fatal runtime exception when the program or function is not found.

We might be able to do this with the Dynamic type, but we need to benchmark how much slower that would be and if it is compatible with NativeAOT. It might have a big performance impact if every program or function pointer has to use the dynamic type to hold their references.

If all else fails, we might have to look into emitting IL into a sort of custom assembly and try to bypass all C# checks, or use another "intermediate" language to hold these signatures and calls for us.

### Community help and suggestions?

We'll leave this issue open for discussion and suggestions until we figure out the best way to implement and handle this functionality. We might need community help on this, from both the COBOL side and the C# side.

**Comment by sgorozco** [comment]
Hello!  
I have written several years ago something that might tackle the challenge that you describe, when I implemented a multiple-dispatch mechanism for C# (more or less at the same time that `dynamic` came out and enabled multi-dispatch natively).

This was the basic strategy that I used:

1. A resolver class was written; its purpose was to return a weakly-typed delegate (all params as type object) that could be used to initiate the multiple-dispatch call with types only known at runtime.  
2. The class receives a method name (provided as a string) as well as an array of types of each concrete parameter that will be passed to the returned delegate).
3. The class then assembles a strong 64-bit signature, based on the combination of `GetHashValue()` calls for every parameter's concrete type.
4. Using this strong signature as key, the class would check an internal cache, implemented as a Dictionary of delegates. If the delegate was found, it was returned immediately to the caller. 
5. If not found, then Reflection was used to perform the method lookup.  Reflection does a perfect job of selecting the most specific method that will accept the parameters we intend to pass; however this is *slow*.
6. Once the method was found, an open delegate was created dynamically for it using the Expression API. This involves assembling programmatically an expression that would look as something like this if written in code:
```
    void M(object instance, params object[] arguments) => {
         // Typecast parameters to perform a static call to the adequate wrapped method
         ((CallingType)instance).Method((TParam0)arguments[0], (TParam1)arguments[1], ...);
     };

```
Basically, it is dynamically compiling a weakly-typed delegate (receiving all parameters as type object), that casts every parameter to its real type, and then invokes the real, strongly-typed method (the one that was resolved via Reflection).

7. The delegate is cached for future use.  Reflection is only used **the very first time** a certain combination of parameters is found. Then it is returned to the caller, so it can proceed to invoke the selected method via this delegate.

I've done benchmarking of this method, and it is around 15% faster than using `dynamic `to perform the multiple-dispatch, with the added flexibility of allowing the method lookup and binding via a `string`. However, both this method and `dynamic `are around 600x faster than calling `Invoke()` on the `MethodInfo `returned by Reflection!  (This was actually measured by John Skeet of Stack Overflow fame).

I have not published this source code anywhere; would you be interested if I make a gist out of it so you can play with it?   (I would paste it tomorrow as it is already midnight for me, but I found a tweet asking for suggestions to this challenge and it rang all bells inside my head) 

 ;)

**Comment by sgorozco** [comment]
Improved formatting

**Comment by xoofx** [comment]
Hello there, coming from this Twitter [comment](https://twitter.com/terrajobst/status/1630411383149047811)

> We're trying to avoid the use of reflection in order to support NativeAOT, with both the compiler and the generated C# code.

Note that NativeAOT does actually support relatively well reflection (including method invoke) with some hints (see below). What it doesn't support is e.g loading dynamic IL executed at runtime or binding a type dynamically without telling NativeAOT that you want to keep this type and code around.

The main requirement you need for NativeAOT with reflection is make sure that types that you are going to access are visible to the compiler and that you are properly adding the attribute [DynamicallyAccessedMembers](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?) to hint about this access, similar to this sample code:

```c#
// Compile with AOT: dotnet publish -c Release -r win-x64 -p:PublishAot=true
using System.Diagnostics.CodeAnalysis;
using System.Reflection;

namespace ConsoleApp1;

internal class Program
{
    private static readonly Dictionary<string, Type> MapNameToType = new();

    static void Main(string[] args)
    {
        // Make sure that Program is accessed explicitly
        RegisterType<Program>("Program");

        // Invoke the type and function by name
        Invoke(MapNameToType["Program"], "Print");
    }

    static void RegisterType<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicMethods | DynamicallyAccessedMemberTypes.PublicMethods)] T>(string name)
    {
        RegisterType(name, typeof(T));
    }

    static void RegisterType(string name, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicMethods | DynamicallyAccessedMemberTypes.PublicMethods)] Type type)
    {
        MapNameToType[name] = type;
    }
    
    static void Invoke([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicMethods | DynamicallyAccessedMemberTypes.PublicMethods)] Type type, string name, params object[] args)
    {
        type.GetMethod(name, BindingFlags.NonPublic | BindingFlags.Static)!.Invoke(null, args);
    }

    static void Print()
    {
        Console.WriteLine("Hello, World!");
    }
}
```

I would avoid any other pseudo custom reflection that would require a lot more work for little benefits, maybe except performance, but even performance for [reflection has been significantly improved in .NET 7](https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/#reflection).

**Comment by GitMensch** [comment]
> It can also be used during runtime when the signature of a program can't be statically checked during compile time

I think this route should not be followed as the external repository is _explicit_ for type checking - so you use it to **actually** _statically check at compile time_.

Also note that with the prototype declaration you can include those definitions directly in the source program (it kind of overrides the external repository for the current implementation - and if you would use whatever you specify to update the repository, then those should be used to update it, too).

**Comment by KTSnowy** [comment]
Hey @GitMensch, in the example below where the program's name is obtained from user input, it can't be statically checked at compile time, right?

Because the value is not known to the compiler, it will have to check it during runtime *after* it receives the user's input.

The standard also defines an exception which is set exist if it can't find it during runtime: "If the runtime system cannot locate the program, the EC-PROGRAM-NOT-FOUND exception condition is set to exist and the value of the address-identifier is the predefined address NULL."

```cobol
*> COBOL allows getting a program pointer from a string:
set ptr-name to address of program str-var.

*> The value of str-var can't be statically checked during 
*> compile time if it is obtained from user input:
accept str-var from standard-input.

*> The name of the program will be checked during runtime
*> and allows calling it afterwards, which will cause another
*> runtime check for its parameters and return types:
call ptr-name using ... returning ... .
```

**Comment by KTSnowy** [comment]
Hey @xoofx, that looks great, thank you! I didn't know about the DynamicallyAccessedMembers attribute and from the page you linked it doesn't seem to have as big of a performance impact in .NET 7 (might be even faster with NativeAOT?)

Would this still work if we use it together with the [ModuleInitializer](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.moduleinitializerattribute?view=net-7.0) attribute? We could keep the External Repository as a separate assembly that is referenced by every project dependency, and each dependency would then register its types through the ModuleInitializer attribute (also checking for duplicate names and conflicts).

I'm unsure how this would work if the External Repository assembly was already compiled previously, and then referenced as a package or through a `<Reference Include="ExternalRepository">`. Would NativeAOT still be able to access the types from project dependencies?

**Comment by GitMensch** [comment]
> Because the value is not known to the compiler, it will have to check it during runtime after it receives the user's input.

yes, but not via repository

> The standard also defines an exception which is set exist if it can't find it during runtime: "If the runtime system cannot locate the program, the EC-PROGRAM-NOT-FOUND exception condition is set to exist and the value of the address-identifier is the predefined address NULL."

That is the exception handling, which is runtime only - and this is just about the program not being there on the `SET` statement.
For the `CALL` it would be a different exception if the pointer is `NULL` and another different exception if the program was called but does not match the call signature.
To check the later the _calling_ program must make the signature used available "somehow" (in the runtime directly, or in an attribute which can be accessed by the caller) and the _called_ program must check it; in this case only if the programs are compiled with the appropriate runtime checks enabled!
To provide better type checking COBOL also allows "typed pointers" - only with these (not sure if there's a COBOL implementation that actually supports them...) a prototype check is done at compile time.

**Comment by KTSnowy** [comment]
> yes, but not via repository

Hmmmm, but that means we'd have to instantiate another data structure to register and hold each of these values, and they would ultimately lead to a signature from the repository anyways to check for the parameters, so this could save a bit of memory in the end (probably?). 

The Standard usually doesn't define anything related to the compilation process itself (Like lexing, parsing or codegen), so it would be odd for it to define a symbol table structure used only during compilation. Does the standard specify where the external repository can or can't be used?

This is not used if Otterkit can statically check and find the correct program at compile time, only if it can't statically check the value it will then attempt at runtime through the repository (which would already hold these values).

> That is the exception handling, which is runtime only - and this is just about the program not being there on the SET statement.
For the CALL it would be a different exception if the pointer is NULL and another different exception if the program was called but does not match the call signature.

@GitMensch I might be a bit confused on how the standard is defining these, please let me know if I'm wrong. 

That exception is from the `address of program` identifier which is not part of the `SET` statement. The identifier allows getting a program pointer from an alphanumeric data item, and if that data item's value is only obtained during runtime from user input then the compiler won't be able to statically check the program's name, because the data item is "empty" at compile-time.

If the `address of program` turns out to not be null after getting user input, we'd still have to check during runtime for the parameters and return item if the pointer is used in a `CALL` statement, which will set the `EC-PROGRAM-ARG-MISMATCH` to exist if it doesn't match what is defined in the signature for that program pointer. In Otterkit's case, the program pointer could hold a reference to a signature from the external repository (a `Signature` type to hold the reference, because C# doesn't allow raw method pointers).

This means that in certain cases there would be two runtime checks, one to find the program's name after getting an alphanumeric from standard-input and another to check for conformance with the parameters (if the pointer is used in a `CALL`). This is not needed if the value passed to the `address of program` identifier is static and known to the compiler during compile time.

Maybe I'm missing something? Please let me know.

> To check the later the calling program must make the signature used available "somehow" (in the runtime directly, or in an attribute which can be accessed by the caller) and the called program must check it; in this case only if the programs are compiled with the appropriate runtime checks enabled!

Wait, do you mean if exception checking is enabled or is there any other kind of checks that the standard allows the user to enable?

> To provide better type checking COBOL also allows "typed pointers" - only with these (not sure if there's a COBOL implementation that actually supports them...) a prototype check is done at compile time.

Otterkit will support typed pointers 😉, it's honestly an awesome feature. Makes it so much easier to statically check and maintain type safety even when using pointers.

C# calls those ["nullable reference types"](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-reference-types), where a variable can only contain a reference to a specific type or null.

**Comment by KTSnowy** [comment]
> I've done benchmarking of this method, and it is around 15% faster than using dynamic to perform the multiple-dispatch, with the added flexibility of allowing the method lookup and binding via a string. However, both this method and dynamic are around 600x faster than calling Invoke() on the MethodInfo returned by Reflection! (This was actually measured by John Skeet of Stack Overflow fame).

Hey @sgorozco, would this still be faster in .NET 7 compared to @xoofx's reflection-based approach?

> I have not published this source code anywhere; would you be interested if I make a gist out of it so you can play with it?

Sure, that would be great! We could then benchmark both approaches to check which one would perform better for Otterkit's purposes.

Does your source code have any specific licensing notice? Otterkit uses the Apache-2.0 license so we won't be able to include any GPL-license code or any code with a license incompatible with the Apache license.

**Comment by GitMensch** [comment]
> Does the standard specify where the external repository can or can't be used?

Not that I remember, you are free to check for yourself :-)

>This means that in certain cases there would be two runtime checks

yes

>This is not needed if the value passed to the address of program identifier is static and known to the compiler during compile time

no, it still is needed:
* you may still not be able to find it and then need to set the exception and the variable to NULL - if you don't have the program's address resolved because the program itself is included in the same compile unit (this is not related to the repositroy)
* you still need to check for `NULL` on the call in any case - because the variable may be changed at runtime afterwards - and set an exception and possibly go into the `ON EXCEPTION` branch because of this
* you still need to check for conformance _even if_ you use a typed pointer on the caller's side because only the caller's side knows where it was called from and if the type actually matches

For the last point: remember that the `CALL`ed program may not be what the repository says, either because the repository is outdated (for example was not updated or the caller was compiled before the callee was adjusted). You therefore nearly always have to check on the callee's side (at least if dynamic called = not statically linked in, when this can be indeed optimized out).

The standard has rules for conformance which apply to CALL and INVOKE, where the repository is used to:
* potentially error during compile time because of mismatch
* potentially setup the call correctly for BY VALUE (if the implementation needs that)

The standard has additional rules for conformance which apply to the callee (so what is happening if the program/function/method could be resolved).

>  I might be a bit confused on how the standard is defining these

The biggest thing to get confused with is the actual exception handling. There are several places where the standard says that an exception is set, but this doesn't also mean that this has any effect.

It _always_ has an effect when there's an `ON EXCEPTION`, `ON INVALID KEY` and similar code, because these code parts have to be taken then.

Otherwise it is only available to the program if checking code is explicit enabled for this specific exception - by default all exception checks are defined to be turned off!

In most cases if there is an exception and you don't have checking for the specific exception enabled then "whatever happens, happens". A `COMPUTE` may raise an `ON SIZE ERROR` if a packed decimal value is all x'FF' or interpret that as zero or as all 9 or whatever, an implementation is even free to crash hard without any more information.

For many exceptions with no check enabled implementations will do something "useful" (and/or what other implementations do). For example many implementations just use the second half-byte for `PIC 9 USAGE DISPLAY` (recent GnuCOBOL does that) so that all of a low value, an EBCDIC SPACE, an ASCII SPACE an ASCII ZERO and an EBCDIC ZERO will be seen as "zero digit".

Back to the CALL example you'd likely reset the complete CALL related exceptions at the start of the actual CALL, then the appropriate exception in the runtime it if the runtime can't resolve the program [can be done inline for CALL ptr-var and ptr-var being NULL], then check in the calling program for "is any call related exception set" to decide which code path to take.
If no explicit check is coded and the runtime checks are not enabled then you'd even be free to just go on (ignore the call), but I don't know of any implementation that doesn't abort here.
So, if the program was resolved (= found, however you do that in your implementation) you can then get into that program and (if the parameter checking is enabled in both the caller and callee!) check for the matching signature, setting (other) call related exceptions to go back to the caller.

By default you would have zero runtime penalty for signature matching, because by default all runtime exceptions are disabled.
You could internally split a CALL to multiple functions on the callee side, for example "resolve" (for CALL literal this would be done likely only once), then the "setup parameters to be passed to the callee", then "check" (if the caller has the check enabled; for CALL literal you likely would also do this only once, for CALL variable possibly only if the value changed, same with the resolve; if the callee hasn't the check enabled this would be called but do nothing), then the actual calling - but of course that's up to you :-)

**Comment by KTSnowy** [comment]
Thank you, this cleared most of the confusion I had! 

> By default you would have zero runtime penalty for signature matching, because by default all runtime exceptions are disabled.

By "default" you mean if we choose to ignore the call in case something goes wrong and exceptions are turned off, right? 

> If no explicit check is coded and the runtime checks are not enabled then you'd even be free to just go on (ignore the call), but I don't know of any implementation that doesn't abort here.

If we choose to abort in case something goes wrong (like the signature not matching) then we'd still need some sort of signature matching in order to check if we should terminate or not, that might still have *some* runtime penalty (but probably not something to worry about).

> For the last point: remember that the CALLed program may not be what the repository says, either because the repository is outdated (for example was not updated or the caller was compiled before the callee was adjusted). You therefore nearly always have to check on the callee's side (at least if dynamic called = not statically linked in, when this can be indeed optimized out).

Maybe we could go with a compile time "hash" of the signature (something like C++'s mangled names?). The compiler would have two "hashes", the caller hash and the callee hash. The callee hash is cached in the program or method code itself as a compiler-generated readonly static property. The caller hash is generated during compile time for the `CALL` and `INVOKE` statements only, and contains an "expected hash" based on the statement's `USING` and `RETURNING` items.

This way we can do signature matching in just one hash comparison during runtime, checking if the "expected signature" in the `CALL` statement matches the actual signature of the program being called.

Doing this would be faster than going through a list of parameters, return, and data item types. Could allow us to quickly check before deciding to attempt a call to a C# method with incorrect number of parameters if we use @xoofx's solution (Should throw a C# exception otherwise, and those are expensive to catch).

Or maybe we could completely bypass C#'s way to passing parameters (if that's possible somehow), and end up just calling a void method with no parameters after those are passed to the COBOL program in a "different" way. If that's possible @xoofx's Invoke method could turn into `.Invoke(null, null);`

Edit: Maybe the hash idea is too complex for this?

**Comment by sgorozco** [comment]
> Hey @sgorozco, would this still be faster in .NET 7 compared to @xoofx's reflection-based approach?

Can't tell, @xoofx's mentions that reflection `Invoke()` performance has been improved in the latest dotNet releases.  The code I'm sharing is quite old. Most probably the current ` Invoke()` beats or matches the performance of my tiny vintage library.

> Does your source code have any specific licensing notice? Otterkit uses the Apache-2.0 license so we won't be able to include any GPL-license code or any code with a license incompatible with the Apache license.

It is being released as Public domain.

[Here is the .net Fiddle where you can give it a try if you like.](https://dotnetfiddle.net/okGtsm)  It is compilable under .net Framework 4.7.2 (in Visual Studio) yet for some reason it doesn't run in dotNet Fiddle unless I specify .Net 7 as the target.

The gist of the library is in the `WrapMethod()` method, where you can see how the dynamic delegate is compiled.  Again, this was how it was done back then, most probably there is an easier way now.  The code provided by @xoofx  is dead-beat simple and elegant for sure!  =)

_Note: I made a slight update to the .netFiddle code, I was missing the GetFuncDelegate() methods and was obtaining a Func<> Delegate via the GetDelegate() method which is supposed to be used to receive an Action() delegate.  This worked because the code is virtually the same, but in the case that the TDelegate parameter had been incompatible with the number of parameters being passed to the resolver, then the diagnostic error message would have been wrong._

**Comment by GitMensch** [comment]
The hash approach is reasonable for a "quick check", if that fails and the caller has ON OVERFLOW, then you can directly go there, if not then you can do a "full" check to raise a nice error message, but this can also be done (much) later.

The initial check for "resolve worked" should also be quick and fast.

**Comment by KTSnowy** [comment]
Just made a new commit to add the [System.IO.Hashing](https://learn.microsoft.com/en-us/dotnet/api/system.io.hashing?view=dotnet-plat-ext-7.0) library, this will allow us to use the fast xxh64 hashing algorithm for the signatures, and when .NET 8 releases we'll be able to use the new xxh3 algorithm.

[These two hashing algorithms are amazing in both performance and collision resistance](https://github.com/Cyan4973/xxHash), so it should be pretty safe for us to use them for the signatures.

I'll post an implementation of the External Repository here soon, based on @xoofx's solution. Just need to test a few things and make a commit for it in the runtime library repo.

**Comment by xoofx** [comment]
> Would this still work if we use it together with the [ModuleInitializer](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.moduleinitializerattribute?view=net-7.0) attribute? We could keep the External Repository as a separate assembly that is referenced by every project dependency, and each dependency would then register its types through the ModuleInitializer attribute (also checking for duplicate names and conflicts).

Yes, `ModuleInitializer` is properly handled by NativeAOT, so anything called transitively from there would be preserved.

**Comment by KTSnowy** [comment]
Here's the current implementation I have based on @xoofx's solution;

This is the Signature class to hold both the compiled COBOL program to C# class and the hash of the signature. It's still missing some required information like the source unit's type: 

```csharp
using System.Diagnostics.CodeAnalysis;

namespace OtterkitLibrary;

public class Signature
{
    [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
    internal Type Type { get; init; }
    internal ReadOnlyMemory<byte> Hash { get; init; }

    public Signature([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type type, ReadOnlyMemory<byte> hash)
    {
        this.Type = type;
        this.Hash = hash;
    }
}
```

And here is the implementation based on @xoofx's solution. There's also a compiled COBOL program here called `"hello-world"` to test the implementation, and a `UsageExample()` method to register and invoke the program.
```csharp
using System.Diagnostics.CodeAnalysis;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;
using System.Reflection;

namespace OtterkitLibrary;

// PROGRAM-ID. hello-world.
public class _hello_world
{
    // WORKING-STORAGE SECTION.
    private static Alphanumeric _hello = new("Hello, World!"u8, 0, 20, new byte[20]);

    // LOCAL-STORAGE SECTION.

    // PROCEDURE DIVISION.
    public static void Procedure()
    {
        // PROCEDURE STATEMENTS.
        Statements.DISPLAY(" ", true, _hello.Display, String.Empty);
        Statements.STOP(false, "\nTerminating Now!");
    }
}

public static partial class External
{
    private static readonly Dictionary<string, Signature> ExternalRepository = new();
    
    public static void UsageExample()
    {
        // var signatureHash = "program:hello-world|params:null|ret:null|ec:null|comma:f|curr:null"u8;
        // var xxhash = XxHash64.Hash(signatureHash, 172575);
        // Result: 9, 232, 166, 24, 201, 31, 204, 132

        var hash = new byte[] { 9, 232, 166, 24, 201, 31, 204, 132 };

        RegisterSignature<_hello_world>("hello-world", hash);

        // Would be equivalent to a COBOL program pointer
        var signature = GetSignature("hello-world");

        // If the program pointer is not null:
        if (signature is not null) Invoke(signature.Type, Array.Empty<ICOBOLType>());

        // If the program pointer is null:
        if (signature is null) ExceptionRegistry.ActivateException("EC-PROGRAM-PTR-NULL");
    }

    public static void RegisterSignature<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] T>(string name, ReadOnlyMemory<byte> hash)
    {
        RegisterSignature(name, typeof(T), hash);
    }

    public static void RegisterSignature(string name, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type type, ReadOnlyMemory<byte> hash)
    {
        ExternalRepository[name] = new Signature(type, hash);
    }

    public static Signature? GetSignature(string name)
    {
        ref var signatureOrNull = ref CollectionsMarshal.GetValueRefOrNullRef(ExternalRepository, name);

        if (Unsafe.IsNullRef(ref signatureOrNull))
        {
            ExceptionRegistry.ActivateException("EC-PROGRAM-NOT-FOUND");
            return null;
        }

        return signatureOrNull;
    }

    public static void Invoke([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type type, params ICOBOLType[] args)
    {
        var method = type.GetMethod("Procedure", BindingFlags.Public | BindingFlags.Static);

        if (method is not null) method.Invoke(null, args);

        if (method is null) ExceptionRegistry.ActivateException("EC-PROGRAM-ARG-MISMATCH");
    }
}
```

The program outputs "Hello, World!" as expected, even when compiled with NativeAOT!

I'm unsure if activating the exceptions should be done here or outside of these methods, @GitMensch what do you think? 

The PR is open here, we might need feedback before merging to main: https://github.com/otterkit/libotterkit/pull/27

**Comment by KTSnowy** [comment]
Another thing that I'm not entirely sure about is the ReadOnlyMemory<byte> to store the hash, maybe there are more efficient alternatives to store the xxh64 hash bytes.

Edit: Actually there is, `BitConverter.ToInt64()` will convert the hash bytes into their Int64 representation! I should update the PR to reflect this.

**Comment by GitMensch** [comment]
I don't know enough about C# to comment much on this, but this doesn't look plain wrong, so... :-)

Depending on how `ActivateException` gets to know where it is called from I'd suggest adding "this.position" or something into that method as it has to store where the last exception was raised (even 10 lines after, the exception will stay "active" until `SET LAST EXCEPTION OFF` (or similar) is used or another exception happens - and in the meantime the program that raised the exception may also be (physical) canceled..

For exceptions at least if you turned them on with locations - you obviously need to store the location somewhere; I'd suggest a "pos" attribute (or similar) which has the current program/function section/paragraph and source (file+line). If location is not turned on then only the program/function would be set.

Note: you could use the same "pos" attribute to implement `READY TRACE`/ `RESET TRACE`.

**Comment by KTSnowy** [comment]
> Depending on how ActivateException gets to know where it is called from I'd suggest adding "this.position" or something into that method as it has to store where the last exception was raised

I'll try to figure out an implementation for it. Currently the `ActivateException` just sets it to true and sets the `LAST EXCEPTION` to the activated exception.

Hmmmm, maybe activating the exceptions should be done outside the Repository methods, so that we can pass the position and location to it without having to also pass those to the `Invoke` and `GetSignature` methods. 

@GitMensch what do you think of the hash format?

```
program:hello-world|params:null|ret:null|ec:null|comma:f|curr:null
```

This string "hash" would be converted at compile time into an actual int64 hash value using the xxHash64 algorithm, which in theory means that we could do the "quick check" on the `CALL` and `INVOKE` statements with a single int64 comparison against the "expected hash", and then fallback to a longer check if that fails, so that we can display a nicer error message like you suggested yesterday. 

If the hash of the callee is the above and the expected hash of the `CALL` is this one:
```
                 DECIMAL-POINT IS COMMA is specified \/  
program:hello-world|params:null|ret:null|ec:null|comma:t|curr:null
```
Because the expected hash is a program with `DECIMAL-POINT IS COMMA` clause, and the actual hash is a program without the clause the quick check would fail due to a different int64 value being produced by the xxHash64 for both "hashes".

The string representation might also be useful for debugging purposes.

Edit: Nulls for when there are no parameters and no return value, but I still need to figure out how to encode type information into the string for when paramaters are present.

**Comment by GitMensch** [comment]
not sure about the format:
* Do you need the program name in there? It definitely would be useful for debugging, but otherwise?
* Do you really want to fail for comma+currency sign?
* You possibly don't want to fail if exceptions don't match. Consider that "old version" raised EC-BOUNDS-SUBSCRIPTS an a new version raises the group EC-BOUNDS - or vice versa.

For the parameters you likely want to encode only:
* amount
* BY VALUE / otherwise
* size for everything not passed BY VALUE
* integer type for BY VALUE
* optionality

You could place all this in an array (size is automatically in) and get something like

```
[] -> no parameters
[5,5156,opt:int64] -> three parameters, first two with size 5 and 5156, the third one an OPTIONAL BY VALUE of maximum size int64
```

if you don't want to limit on fixed integer sizes but digit sizes use V38 (amount of maximum digits) instead.

**Comment by KTSnowy** [comment]
Hey @GitMensch, I just checked the standard again and you're right, we shouldn't fail if exceptions don't match and we don't need the program name there (maybe only if a "debug build" is on?).

But in certain situations we should still fail for the comma and currency sign when the parameter is elementary and passed by reference. The standard specifies that it should fail if the element being called is a program with a program-specifier in the repository paragraph of the caller, a program and the `NESTED` phrase is used on the `CALL`, a class method or a function.

> then the definition of the formal parameter and the definition of the argument shall have the same ALIGN, BLANK WHEN ZERO, DYNAMIC LENGTH, JUSTIFIED, PICTURE, SIGN, and USAGE clauses, with the following exceptions:
> - Currency symbols match if and only if the corresponding currency strings are the same.
> - Period picture symbols match if and only if the DECIMAL-POINT IS COMMA clause is in effect for both the activating and the activated runtime elements or for neither of them. Comma picture symbols match if and only if the DECIMAL-POINT IS COMMA clause is in effect for both the activating and the activated runtime elements or for neither of them.

Otherwise if the called element is a program without a program-specifier in the repository paragraph of the caller and without a `NESTED` in the `CALL` statement then just the item sizes are checked and the array format would work just fine.

But we still need to figure out how to encode this extra information into it somehow for the other source element types and for when the program is specified in the caller's repository.

I'm not sure how to encode all of that in a way that is still performant when runtime checks are needed, that is a lot of stuff to check. Please let me know if you have any suggestions, maybe we could hash just the "type" information instead of the whole thing.

**Comment by KTSnowy** [comment]
Maybe some of this information could be statically stored in a field or property inside the items themselves.

So that a picture check becomes just `_ws_param.Picture == _ws_arg.Picture`.

**Comment by KTSnowy** [comment]
The PR (https://github.com/otterkit/libotterkit/pull/27) was merged, so I'm closing this issue for now. 

If anyone finds an issue with the merged implementation please let me know, and I'll reopen this issue to discuss it a bit more.
